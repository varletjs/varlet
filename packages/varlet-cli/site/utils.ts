import { get } from 'lodash-es'
import { StyleProvider, type Theme, withSiteConfigNamespace } from '@varlet/cli/client'

export interface Menu {
  doc: string
  text: Record<string, string>
  type: MenuTypes
  useMobile?: boolean
}

export enum MenuTypes {
  TITLE = 1,
  COMPONENT = 2,
  DOCUMENTATION = 3,
}

export function removeEmpty(object: Record<string, string> = {}) {
  return Object.keys(object).reduce((record: Record<string, string>, key) => {
    const value = object[key]
    value && (record[key] = value)
    return record
  }, {})
}

export function isPhone() {
  return /Android|webOS|iPhone|iPod|BlackBerry|Pad/i.test(navigator.userAgent)
}

export function inIframe() {
  return window.self !== window.top
}

export function setTheme(config: Record<string, any>, name: Theme) {
  const styleVars = withSiteConfigNamespace(get(config, name, {}))

  StyleProvider(styleVars)
}

export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}
