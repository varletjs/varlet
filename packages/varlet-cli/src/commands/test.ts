import execa from 'execa'
import { VITEST_CONFIG } from '../shared/constant'
import { buildSiteEntry } from '../compiler/compileSiteEntry'

interface TestCommandOptions {
  watch?: boolean
  component?: string
  coverage?: boolean
}

export async function test({ component, watch, coverage }: TestCommandOptions) {
  process.env.NODE_ENV = 'test'

  // Test cases may contain documentation example code, and site dependencies need to build
  await buildSiteEntry()

  const args = ['--config', VITEST_CONFIG]

  if (!watch) {
    args.unshift('run')
  }

  if (coverage) {
    args.push('--coverage')
  }

  if (component) {
    args.push('--dir', `src/${component.trim()}`)
  }

  await execa('vitest', args, { stdin: 'inherit', stdout: 'inherit', stderr: 'inherit' })
}
