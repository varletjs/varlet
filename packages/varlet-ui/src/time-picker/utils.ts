import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
