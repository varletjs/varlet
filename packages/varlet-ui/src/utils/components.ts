import {
  createApp,
  h,
  getCurrentInstance,
  inject,
  onUnmounted,
  computed,
  provide,
  reactive,
  isVNode,
  onMounted,
  onBeforeUnmount,
  nextTick,
  ref,
  onActivated,
  onDeactivated,
} from 'vue'
import type { Component, VNode, ComputedRef, ComponentInternalInstance, Ref, ComponentPublicInstance } from 'vue'
import { isArray, removeItem } from '@varlet/shared'

export interface MountInstance {
  instance: ComponentPublicInstance
  unmount: () => void
}

export interface ChildrenCounter {
  collect(instance: ComponentInternalInstance): void

  clear(instance: ComponentInternalInstance): void

  instances: ComponentInternalInstance[]
}

export interface BaseParentProvider<C> {
  collect(childProvider: C): void

  clear(childProvider: C): void
}

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export function mount(component: Component): MountInstance {
  const app = createApp(component)
  const host = document.createElement('div')
  document.body.appendChild(host)

  return {
    instance: app.mount(host),
    unmount() {
      app.unmount()
      document.body.removeChild(host)
    },
  }
}

export function mountInstance(
  component: any,
  props: Record<string, any> = {},
  eventListener: Record<string, any> = {}
): {
  unmountInstance: () => void
} {
  const Host = {
    setup() {
      return () =>
        h(component, {
          ...props,
          ...eventListener,
        })
    },
  }

  const { unmount } = mount(Host)
  return { unmountInstance: unmount }
}

export function flatVNodes(subTree: any) {
  const vNodes: VNode[] = []

  const flat = (subTree: any) => {
    if (subTree?.component) {
      flat(subTree?.component.subTree)
      return
    }

    if (Array.isArray(subTree?.children)) {
      subTree.children.forEach((child: any) => {
        if (isVNode(child)) {
          vNodes.push(child)

          flat(child)
        }
      })
    }
  }

  flat(subTree)

  return vNodes
}

export function useAtChildrenCounter(key: symbol) {
  const instances: ComponentInternalInstance[] = reactive([])
  const parentInstance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  const sortInstances = () => {
    const vNodes: any[] = flatVNodes(parentInstance.subTree)

    instances.sort((a, b) => {
      return vNodes.indexOf(a.vnode) - vNodes.indexOf(b.vnode)
    })
  }

  const collect = (instance: ComponentInternalInstance) => {
    instances.push(instance)
    sortInstances()
  }

  const clear = (instance: ComponentInternalInstance) => {
    removeItem(instances, instance)
  }

  provide<ChildrenCounter>(key, {
    collect,
    clear,
    instances,
  })

  const length: ComputedRef<number> = computed(() => instances.length)

  return {
    length,
  }
}

export function useAtParentIndex(key: symbol) {
  if (!keyInProvides(key)) {
    return { index: null }
  }

  const childrenCounter: ChildrenCounter = inject<ChildrenCounter>(key) as ChildrenCounter

  const { collect, clear, instances } = childrenCounter

  const instance: ComponentInternalInstance = getCurrentInstance() as ComponentInternalInstance

  onMounted(() => {
    nextTick().then(() => collect(instance))
  })
  onUnmounted(() => {
    nextTick().then(() => clear(instance))
  })

  const index = computed(() => instances.indexOf(instance))

  return {
    index,
  }
}

export function useChildren<P, C>(key: symbol) {
  const childProviders: C[] = []

  const collect = (childProvider: C) => {
    childProviders.push(childProvider)
  }

  const clear = (childProvider: C) => {
    removeItem(childProviders, childProvider)
  }

  const bindChildren = (parentProvider: P) => {
    provide<P & BaseParentProvider<C>>(key, {
      collect,
      clear,
      ...parentProvider,
    })
  }

  return {
    childProviders,
    bindChildren,
  }
}

export function useParent<P, C>(key: symbol) {
  if (!keyInProvides(key)) {
    return {
      parentProvider: null,
      bindParent: null,
    }
  }

  const rawParentProvider = inject<P & BaseParentProvider<C>>(key) as P & BaseParentProvider<C>

  const { collect, clear, ...parentProvider } = rawParentProvider

  const bindParent = (childProvider: C) => {
    onMounted(() => collect(childProvider))
    onBeforeUnmount(() => clear(childProvider))
  }

  return {
    parentProvider,
    bindParent,
  }
}

export function keyInProvides(key: symbol) {
  const instance = getCurrentInstance() as any

  return key in instance.provides
}

export function useValidation() {
  const errorMessage: Ref<string> = ref('')

  const validate = async (rules: any, value: any, apis?: any): Promise<boolean> => {
    if (!isArray(rules) || !rules.length) {
      return true
    }

    const resArr = await Promise.all(rules.map((rule) => rule(value, apis)))

    return !resArr.some((res) => {
      if (res !== true) {
        errorMessage.value = String(res)
        return true
      }

      return false
    })
  }

  const resetValidation = () => {
    errorMessage.value = ''
  }

  const validateWithTrigger = async <T>(validateTrigger: T[], trigger: T, rules: any, value: any, apis?: any) => {
    if (validateTrigger.includes(trigger)) {
      ;(await validate(rules, value, apis)) && (errorMessage.value = '')
    }
  }

  return {
    errorMessage,
    validate,
    resetValidation,
    validateWithTrigger,
  }
}

export function useRouteListener(cb: () => void) {
  onMounted(() => {
    window.addEventListener('hashchange', cb)
    window.addEventListener('popstate', cb)
  })
  onUnmounted(() => {
    window.removeEventListener('hashchange', cb)
    window.removeEventListener('popstate', cb)
  })
}

export function useTeleport() {
  const disabled: Ref<boolean> = ref(false)

  onActivated(() => {
    disabled.value = false
  })

  onDeactivated(() => {
    disabled.value = true
  })

  return {
    disabled,
  }
}

export function exposeApis<T = Record<string, any>>(apis: T) {
  const instance = getCurrentInstance()
  if (instance) {
    Object.assign(instance.proxy!, apis)
  }
}

type ClassName = string | undefined | null
type Classes = (ClassName | [any, ClassName, ClassName?])[]

export function createNamespace(name: string) {
  const namespace = `var-${name}`

  const createBEM = (suffix?: string): string => {
    if (!suffix) return namespace

    return suffix.startsWith('--') ? `${namespace}${suffix}` : `${namespace}__${suffix}`
  }

  const classes = (...classes: Classes): any[] => {
    return classes.map((className) => {
      if (isArray(className)) {
        const [condition, truthy, falsy = null] = className
        return condition ? truthy : falsy
      }

      return className
    })
  }

  return {
    n: createBEM,
    classes,
  }
}

export function call<F extends (...arg: any) => any, P extends Parameters<F>>(
  fn?: F | null,
  ...arg: P
): ReturnType<F> | undefined {
  if (fn) return fn(...arg)
}
