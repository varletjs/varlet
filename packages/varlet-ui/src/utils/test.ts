import { nextTick } from 'vue'
import type { ComponentPublicInstance } from 'vue'
import { config } from '@vue/test-utils'
import type { VueWrapper, DOMWrapper } from '@vue/test-utils'
import { isPlainObject } from '@varlet/shared'

export const delay = (time: number) => new Promise((resolve) => setTimeout(resolve, time))

export function getTouch(el: Element | Window, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function mockDoubleRaf() {
  const originMethod = window.requestAnimationFrame

  Object.assign(window, {
    requestAnimationFrame(fn: any) {
      setTimeout(fn, 16)
    },
  })

  return {
    mockRestore() {
      window.requestAnimationFrame = originMethod
    },
  }
}

export function trigger(
  wrapper: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | Element | Window,
  eventName: string,
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = new CustomEvent(eventName, { bubbles: true, cancelable: true, detail: {} })

  Object.assign(event, {
    clientX: x,
    clientY: y,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  })

  el.dispatchEvent(event)

  return nextTick()
}

export function mockOffset({
  offsetWidth,
  offsetHeight,
  clientWidth,
  clientHeight,
  offsetLeft,
  offsetTop,
}: {
  offsetWidth?: number
  offsetHeight?: number
  clientWidth?: number
  clientHeight?: number
  offsetLeft?: number
  offsetTop?: number
} = {}) {
  Object.defineProperties(HTMLElement.prototype, {
    offsetWidth: {
      get() {
        return offsetWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    offsetHeight: {
      get() {
        return offsetHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    clientWidth: {
      get() {
        return clientWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    clientHeight: {
      get() {
        return clientHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    offsetLeft: {
      get() {
        return offsetLeft ?? (parseFloat(window.getComputedStyle(this).marginLeft) || 0)
      },
    },
    offsetTop: {
      get() {
        return offsetTop ?? (parseFloat(window.getComputedStyle(this).marginTop) || 0)
      },
    },
    offsetParent: {
      get() {
        return this.parentNode ?? {}
      },
    },
  })
}

export function mockImageNaturalSize(naturalWidth: number, naturalHeight: number) {
  Object.defineProperties(HTMLImageElement.prototype, {
    naturalWidth: {
      get() {
        return naturalWidth
      },
    },
    naturalHeight: {
      get() {
        return naturalHeight
      },
    },
  })
}

export async function triggerDrag(
  el: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | HTMLElement,
  x = 0,
  y = 0
) {
  await trigger(el, 'touchstart', 0, 0)
  await trigger(el, 'touchmove', x / 4, y / 4)
  await trigger(el, 'touchmove', x / 3, y / 3)
  await trigger(el, 'touchmove', x / 2, y / 2)
  await trigger(el, 'touchmove', x, y)
  await trigger(el, 'touchend', x, y)
}

export function mockTranslate() {
  const originMethod = window.getComputedStyle

  const XRE = /translateX\((\d+)px\)/
  const YRE = /translateY\((\d+)px\)/

  Object.assign(window, {
    getComputedStyle: (el: HTMLElement) => {
      const styles = originMethod.call(window, el)
      const x = styles.transform.match(XRE)?.[1] ?? 0
      const y = styles.transform.match(YRE)?.[1] ?? 0
      styles.transform = `matrix(1, 0, 0, 1, ${x}, ${y})`
      return styles
    },
  })

  return {
    mockRestore() {
      window.getComputedStyle = originMethod
    },
  }
}

export function mockFileReader(url: string) {
  const originMethod = window.FileReader

  class FileReader {
    result = ''

    onload() {}

    readAsDataURL() {
      this.result = url
      this.onload()
    }
  }

  Object.assign(window, {
    FileReader,
  })

  return {
    mockRestore() {
      window.FileReader = originMethod
    },
  }
}

export function mockStubs() {
  const originStubs = config.global.stubs

  config.global.stubs = {}

  return {
    mockRestore() {
      config.global.stubs = originStubs
    },
  }
}

export function mockConsole(method: keyof Console, fn: any = () => {}) {
  const originMethod = console[method]

  console[method] = fn

  return {
    mockRestore() {
      console[method] = originMethod as any
    },
  }
}

export function mockScrollTo(Element: any) {
  Element.prototype.scrollTo = function (x: ScrollToOptions | number, y?: number) {
    if (isPlainObject(x)) {
      this.scrollLeft = x.left
      this.scrollTop = x.top
    } else {
      this.scrollLeft = x
      this.scrollTop = y
    }
  }
}
