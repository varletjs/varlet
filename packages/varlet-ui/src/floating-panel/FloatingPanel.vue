<template>
  <Teleport :to="teleport === false ? undefined : teleport" :disabled="teleportDisabled || teleport === false">
    <div
      :class="classes(n(), [safeArea, n('--safe-area')], formatElevation(elevation, 3))"
      :style="{
        height: `${toSizeUnit(maxAnchor)}`,
        transform: `translateY(calc(100% - ${toSizeUnit(visibleHeight)}))`,
        transition: touching
          ? 'none'
          : `transform ${toNumber(
              duration
            )}ms var(--floating-panel-transition-timing-function), background-color 0.25s`,
      }"
      @touchstart="handleTouchstart"
      @touchmove="handleTouchmove"
      @touchend="handleTouchend"
      @touchcancel="handleTouchend"
    >
      <div :class="n('header')">
        <div :class="n('header-toolbar')"></div>
      </div>
      <div :class="n('content')" ref="contentRef">
        <slot></slot>
      </div>
    </div>
  </Teleport>
</template>

<script lang="ts">
import { defineComponent, ref, computed, watch } from 'vue'
import { props } from './props'
import { useLock } from '../context/lock'
import { createNamespace, formatElevation, useTeleport } from '../utils/components'
import { toSizeUnit } from '../utils/elements'
import { useTouch, useWindowSize, useVModel } from '@varlet/use'
import { toNumber, isEmpty, preventDefault, call } from '@varlet/shared'

const { name, n, classes } = createNamespace('floating-panel')

const DEFAULT_START_ANCHOR = 100
const OVERFLOW_REDUCE_RATIO = 0.2

export default defineComponent({
  name,
  props,
  setup(props) {
    const visibleHeight = ref<number>(0)
    const contentRef = ref<HTMLElement | null>(null)
    const { height: windowHeight } = useWindowSize()
    const defaultEndAnchor = computed(() => windowHeight.value * 0.6)
    const anchor = useVModel(props, 'anchor')
    const anchors = computed<number[]>(() => {
      const defaultAnchors = [DEFAULT_START_ANCHOR, defaultEndAnchor.value]
      const { anchors } = props
      return isEmpty(anchors) ? defaultAnchors : anchors!
    })
    const minAnchor = computed<number>(() => Math.min(...anchors.value))
    const maxAnchor = computed<number>(() => Math.max(...anchors.value))
    const { disabled: teleportDisabled } = useTeleport()
    const { deltaY, touching, startTouch, moveTouch, endTouch, isReachTop, isReachBottom } = useTouch()

    let startVisibleHeight: number

    useLock(() => touching.value)

    watch(() => anchor.value, matchAnchor, { immediate: true })

    watch(
      () => anchors.value,
      () => {
        matchAnchor(anchor.value)
      },
      { immediate: true }
    )

    function matchAnchor(anchor: number | undefined | null) {
      setVisibleHeight(anchor != null ? anchor : minAnchor.value)
    }

    function handleTouchstart(event: TouchEvent) {
      startTouch(event)

      startVisibleHeight = visibleHeight.value
    }

    function handleTouchmove(event: TouchEvent) {
      moveTouch(event)

      const target = event.target as Element
      const eventFromContent = contentRef.value === target || contentRef.value?.contains(target)

      if (eventFromContent && !props.contentDraggable) {
        return
      }

      if (
        eventFromContent &&
        props.contentDraggable &&
        visibleHeight.value >= maxAnchor.value &&
        !isReachTop(contentRef.value!)
      ) {
        if (isReachBottom(contentRef.value!)) {
          preventDefault(event)
        }
        return
      }

      const targetVisibleHeight = startVisibleHeight - deltaY.value
      setVisibleHeight(clampVisibleHeight(targetVisibleHeight))
      preventDefault(event)
    }

    function handleTouchend() {
      endTouch()

      const oldAnchor = anchor.value
      setVisibleHeight(visibleHeight.value)
      anchor.value = visibleHeight.value

      if (anchor.value !== oldAnchor) {
        call(props.onAnchorChange, visibleHeight.value)
      }
    }

    function setVisibleHeight(targetVisibleHeight: number) {
      visibleHeight.value = touching.value ? targetVisibleHeight : findNearestAnchor(targetVisibleHeight)
    }

    function clampVisibleHeight(visibleHeight: number): number {
      if (visibleHeight > maxAnchor.value) {
        // overflow top
        const overflowHeight = visibleHeight - maxAnchor.value
        return maxAnchor.value + overflowHeight * OVERFLOW_REDUCE_RATIO
      }

      if (visibleHeight < minAnchor.value) {
        // overflow bottom
        const overflowHeight = minAnchor.value - visibleHeight
        return minAnchor.value - overflowHeight * OVERFLOW_REDUCE_RATIO
      }

      return visibleHeight
    }

    function findNearestAnchor(targetAnchor: number): number {
      if (anchors.value.includes(targetAnchor)) {
        return targetAnchor
      }

      let minDifference = Infinity
      let nearestAnchor = 0

      anchors.value.forEach((anchor) => {
        const difference = Math.abs(anchor - targetAnchor)
        if (difference < minDifference) {
          minDifference = difference
          nearestAnchor = anchor
        }
      })

      return nearestAnchor
    }

    return {
      contentRef,
      teleportDisabled,
      touching,
      minAnchor,
      maxAnchor,
      visibleHeight,
      n,
      classes,
      toSizeUnit,
      toNumber,
      formatElevation,
      handleTouchstart,
      handleTouchmove,
      handleTouchend,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import './floatingPanel.less';
</style>
