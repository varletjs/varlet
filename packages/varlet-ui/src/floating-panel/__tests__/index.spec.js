import FloatingPanel from '..'
import VarFloatingPanel from '../FloatingPanel.vue'
import { createApp } from 'vue'
import { mount } from '@vue/test-utils'
import { describe, expect, vi } from 'vitest'
import { delay, triggerDrag } from '../../utils/test'

test('test floating-panel plugin', () => {
  const app = createApp({}).use(FloatingPanel)
  expect(app.component(FloatingPanel.name)).toBeTruthy()
})

describe('test floating-panel components props', () => {
  test('test floating-panel anchor', async () => {
    const onUpdateAnchor = vi.fn()
    const root = document.createElement('div')
    const wrapper = mount(VarFloatingPanel, {
      props: {
        anchor: 100,
        anchors: [100, 200],
        teleport: root,
        'onUpdate:anchor': onUpdateAnchor,
      },
    })

    const panel = root.querySelector('.var-floating-panel')

    await triggerDrag(panel, 0, 50)
    expect(onUpdateAnchor).toHaveBeenCalledTimes(0)
    expect(panel.outerHTML).toMatchSnapshot()

    await triggerDrag(panel, 0, -200)
    expect(onUpdateAnchor).toHaveBeenCalledTimes(1)
    expect(onUpdateAnchor).toHaveBeenCalledWith(200)
    expect(panel.outerHTML).toMatchSnapshot()

    await triggerDrag(panel, 0, 100)
    expect(onUpdateAnchor).toHaveBeenCalledTimes(2)
    expect(onUpdateAnchor).toHaveBeenCalledWith(100)
    expect(panel.outerHTML).toMatchSnapshot()

    await triggerDrag(panel, 0, -80)
    expect(onUpdateAnchor).toHaveBeenCalledTimes(3)
    expect(onUpdateAnchor).toHaveBeenCalledWith(200)
    expect(panel.outerHTML).toMatchSnapshot()

    wrapper.unmount()
  })

  test('test floating-panel anchors changed', async () => {
    const root = document.createElement('div')
    const wrapper = mount(VarFloatingPanel, {
      props: {
        anchor: 100,
        anchors: [100, 200],
        teleport: root,
      },
    })

    const panel = root.querySelector('.var-floating-panel')

    expect(panel.outerHTML).toMatchSnapshot()

    await wrapper.setProps({
      anchors: [50, 200],
    })

    expect(panel.outerHTML).toMatchSnapshot()
    wrapper.unmount()
  })

  test('test floating-panel contentDraggable', async () => {
    const root = document.createElement('div')
    const wrapper = mount(VarFloatingPanel, {
      props: {
        anchor: 100,
        anchors: [100, 200],
        contentDraggable: true,
        teleport: root,
      },
    })

    const panel = root.querySelector('.var-floating-panel')
    const content = root.querySelector('.var-floating-panel__content')

    await triggerDrag(content, 0, 50)
    expect(panel.outerHTML).toMatchSnapshot()

    await triggerDrag(content, 0, -100)
    expect(panel.outerHTML).toMatchSnapshot()

    await wrapper.setProps({ contentDraggable: false })
    await triggerDrag(content, 0, 100)
    expect(panel.outerHTML).toMatchSnapshot()

    wrapper.unmount()
  })

  test('test floating-panel safeArea', async () => {
    const root = document.createElement('div')
    const wrapper = mount(VarFloatingPanel, {
      props: {
        safeArea: true,
        teleport: root,
      },
    })

    expect(root.querySelector('.var-floating-panel--safe-area')).toBeTruthy()
    await wrapper.setProps({
      safeArea: false,
    })
    expect(root.querySelector('.var-floating-panel--safe-area')).toBeFalsy()
    wrapper.unmount()
  })

  test('test floating-panel elevation', async () => {
    const root = document.createElement('div')
    const wrapper = mount(VarFloatingPanel, {
      props: {
        elevation: true,
        teleport: root,
      },
    })

    expect(root.querySelector('.var-floating-panel').classList.contains('var-elevation--3')).toBe(true)

    await wrapper.setProps({
      elevation: false,
    })

    expect(root.querySelector('.var-floating-panel').classList.contains('var-elevation--3')).toBe(false)

    wrapper.unmount()
  })

  test('test floating-panel duration', async () => {
    const wrapper = mount(VarFloatingPanel, {
      props: {
        duration: 0.4,
      },
    })

    expect(document.body).toMatchSnapshot()

    wrapper.unmount()
  })
})

describe('test floating-panel component slots', () => {
  test('test floating-panel component default', async () => {
    const root = document.createElement('div')
    const wrapper = mount(VarFloatingPanel, {
      props: {
        teleport: root,
      },
      slots: { default: () => 'Hello Varlet!' },
    })

    expect(root.innerHTML).toMatchSnapshot()

    wrapper.unmount()
  })
})

describe('test floating-panel component events', () => {
  test('test floating-panel component onAnchorChange', async () => {
    const onAnchorChange = vi.fn()
    const root = document.createElement('div')
    const wrapper = mount(VarFloatingPanel, {
      props: {
        anchor: 100,
        anchors: [100, 200],
        teleport: root,
        onAnchorChange,
      },
    })

    const panel = root.querySelector('.var-floating-panel')

    await triggerDrag(panel, 0, 50)
    expect(onAnchorChange).toHaveBeenCalledTimes(0)

    await triggerDrag(panel, 0, -200)
    expect(onAnchorChange).toHaveBeenCalledTimes(1)
    expect(onAnchorChange).toHaveBeenCalledWith(200)

    await triggerDrag(panel, 0, 100)
    expect(onAnchorChange).toHaveBeenCalledTimes(2)
    expect(onAnchorChange).toHaveBeenCalledWith(100)

    await triggerDrag(panel, 0, -80)
    expect(onAnchorChange).toHaveBeenCalledTimes(3)
    expect(onAnchorChange).toHaveBeenCalledWith(200)

    wrapper.unmount()
  })
})
