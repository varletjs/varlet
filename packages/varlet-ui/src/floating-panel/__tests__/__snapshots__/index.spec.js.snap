// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test floating-panel component slots > test floating-panel component default 1`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 100px; transform: translateY(calc(100% - 0px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\">Hello Varlet!</div></div>"`;

exports[`test floating-panel components props > test floating-panel anchor 1`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 200px; transform: translateY(calc(100% - 100px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\"></div></div>"`;

exports[`test floating-panel components props > test floating-panel anchor 2`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 200px; transform: translateY(calc(100% - 200px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\"></div></div>"`;

exports[`test floating-panel components props > test floating-panel anchor 3`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 200px; transform: translateY(calc(100% - 100px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\"></div></div>"`;

exports[`test floating-panel components props > test floating-panel anchor 4`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 200px; transform: translateY(calc(100% - 200px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\"></div></div>"`;

exports[`test floating-panel components props > test floating-panel anchors changed 1`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 200px; transform: translateY(calc(100% - 100px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\"></div></div>"`;

exports[`test floating-panel components props > test floating-panel anchors changed 2`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 200px; transform: translateY(calc(100% - 50px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\"></div></div>"`;

exports[`test floating-panel components props > test floating-panel contentDraggable 1`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 200px; transform: translateY(calc(100% - 100px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\"></div></div>"`;

exports[`test floating-panel components props > test floating-panel contentDraggable 2`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 200px; transform: translateY(calc(100% - 200px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\"></div></div>"`;

exports[`test floating-panel components props > test floating-panel contentDraggable 3`] = `"<div class=\\"var-floating-panel var-elevation--2\\" style=\\"height: 200px; transform: translateY(calc(100% - 200px)); transition: transform 300ms var(--floating-panel-transition-timing-function);\\"><div class=\\"var-floating-panel__header\\"><div class=\\"var-floating-panel__header-toolbar\\"></div></div><div class=\\"var-floating-panel__content\\"></div></div>"`;

exports[`test floating-panel components props > test floating-panel duration 1`] = `
<body
  class=""
>
  <div
    class="var-floating-panel var-elevation--2"
    style="height: 100px; transform: translateY(calc(100% - 0px)); transition: transform 0.4ms var(--floating-panel-transition-timing-function);"
  >
    <div
      class="var-floating-panel__header"
    >
      <div
        class="var-floating-panel__header-toolbar"
      />
    </div>
    <div
      class="var-floating-panel__content"
    >
      
      
    </div>
  </div>
  
</body>
`;
